### ISSUE TITLE
Message: Async test method must have Task<T> return type when a result is expected

### ISSUE THREAD

**Issue by mikkelbu** [nunit/nunit.analyzers#49]
We should warn about the following problem.

**Message: Async test method must have Task\<T\> return type when a result is expected**
```csharp
    [TestCase(ExpectedResult = 1)]
    public async System.Threading.Tasks.Task AsyncTaskTestCaseWithExpectedResult()
    {
      await Task.Run(() => 1);
    }
```
and
```csharp
    [TestCase(ExpectedResult = 1)]
    public System.Threading.Tasks.Task TaskTestCaseWithExpectedResult()
    {
      return Task.Run(() => 1);
    }
```

**Comment by jnm2** [comment]
I'd say "Must return a type which produces a result when awaited" when I finish support for ValueTask etc.

**Comment by mikkelbu** [comment]
Or to be precise - and stealing from the [documentation](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/async-return-types), since I have not used async (and especially ValueTask) a lot - "must return a type that has an accessible `GetAwaiter` method. The object returned by the `GetAwaiter` method must implement the `System.Runtime.CompilerServices.ICriticalNotifyCompletion` interface."

**Comment by jnm2** [comment]
@mikkelbu You might want to say something about the type producing a result when awaited though. That description would also fit the non-generic `Task`.

**Comment by mikkelbu** [comment]
@jnm2 You are right. Can one always assume that any type except for the non-generic `Task` which satisfies the above sentence will return a result - or how do one check this?

**Comment by jnm2** [comment]
@mikkelbu The result type of the await expression is always the return type of `.GetAwaiter().GetResult()`. It's void if there is no result.

https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/expressions#awaitable-expressions
